File: pml-cpp/src/pml/sprites/components/hair.cpp
// ═══════════════════════════════════════════════════════════════════════════════
// PML Hair Component — Implementation
// ═══════════════════════════════════════════════════════════════════════════════

#include "hair.h"

#include <algorithm>
#include <cmath>
#include <vector>

namespace pml {

namespace {

using HairParts = std::vector<GraphicObject>;

constexpr double k_head_w = 56.0;

[[nodiscard]] Value poly(const std::vector<double>& coords) {
    std::vector<Value> values;
    values.reserve(coords.size());
    for (double c : coords) {
        values.emplace_back(c);
    }
    return make_list_value(std::move(values));
}

HairParts bangs(const std::string& color) {
    HairParts parts;
    double bw = k_head_w * 0.85;
    double bh = 14.0;
    parts.emplace_back(
        "rect",
        std::unordered_map<std::string, Value>{{"x", -bw / 2.0}, {"y", -k_head_w / 2.0 - 4.0}, {"w", bw}, {"h", bh}},
        color,
        "#1a1a1a",
        1.5);
    double strand_spacing = bw / 6.0;
    for (int i = 0; i < 5; ++i) {
        double sx = -bw / 2.0 + strand_spacing * (i + 1);
        parts.emplace_back(
            "line",
            std::unordered_map<std::string, Value>{{"x1", sx}, {"y1", -k_head_w / 2.0 - 4.0}, {"x2", sx + 3.0}, {"y2", -k_head_w / 2.0 - 4.0 + bh}},
            std::nullopt,
            "#1a1a1a",
            0.8);
    }
    return parts;
}

HairParts create_short(const std::string& color, double /*length*/, bool with_bangs, bool /*highlights*/) {
    HairParts parts;
    double cap_h = 16.0;
    parts.emplace_back(
        "ellipse",
        std::unordered_map<std::string, Value>{{"cx", 0.0}, {"cy", -k_head_w / 2.0 + 4.0}, {"rx", k_head_w / 2.0 + 4.0}, {"ry", cap_h}},
        color,
        "#1a1a1a",
        1.5);
    for (double dx : {-k_head_w / 2.0 - 2.0, k_head_w / 2.0 + 2.0}) {
        parts.emplace_back(
            "ellipse",
            std::unordered_map<std::string, Value>{{"cx", dx}, {"cy", -k_head_w / 4.0}, {"rx", 6.0}, {"ry", 12.0}},
            color,
            "#1a1a1a",
            1.0);
    }
    if (with_bangs) {
        auto b = bangs(color);
        parts.insert(parts.end(), b.begin(), b.end());
    }
    return parts;
}

HairParts create_medium(const std::string& color, double length, bool with_bangs, bool /*highlights*/) {
    HairParts parts;
    parts.emplace_back(
        "ellipse",
        std::unordered_map<std::string, Value>{{"cx", 0.0}, {"cy", -k_head_w / 2.0 + 4.0}, {"rx", k_head_w / 2.0 + 5.0}, {"ry", 18.0}},
        color,
        "#1a1a1a",
        1.5);
    double side_h = std::min(length * 0.6, 40.0);
    for (double dx : {-k_head_w / 2.0 - 4.0, k_head_w / 2.0 + 4.0}) {
        parts.emplace_back(
            "rect",
            std::unordered_map<std::string, Value>{{"x", dx - 8.0}, {"y", -k_head_w / 4.0}, {"w", 16.0}, {"h", side_h}},
            color,
            "#1a1a1a",
            1.0);
    }
    if (with_bangs) {
        auto b = bangs(color);
        parts.insert(parts.end(), b.begin(), b.end());
    }
    return parts;
}

HairParts create_long(const std::string& color, double length, bool with_bangs, bool /*highlights*/) {
    HairParts parts;
    parts.emplace_back(
        "ellipse",
        std::unordered_map<std::string, Value>{{"cx", 0.0}, {"cy", -k_head_w / 2.0 + 4.0}, {"rx", k_head_w / 2.0 + 6.0}, {"ry", 20.0}},
        color,
        "#1a1a1a",
        1.5);
    double side_h = std::min(length * 0.8, 80.0);
    for (double dx : {-k_head_w / 2.0 - 5.0, k_head_w / 2.0 + 5.0}) {
        parts.emplace_back(
            "rect",
            std::unordered_map<std::string, Value>{{"x", dx - 10.0}, {"y", -k_head_w / 4.0}, {"w", 20.0}, {"h", side_h}},
            color,
            "#1a1a1a",
            1.0);
    }
    parts.emplace_back(
        "rect",
        std::unordered_map<std::string, Value>{{"x", -k_head_w / 2.0 - 4.0}, {"y", -k_head_w / 2.0 + 10.0}, {"w", k_head_w + 8.0}, {"h", length * 0.7}},
        color,
        "#1a1a1a",
        1.0);
    if (with_bangs) {
        auto b = bangs(color);
        parts.insert(parts.end(), b.begin(), b.end());
    }
    return parts;
}

HairParts create_ponytail(const std::string& color, double length, bool with_bangs, bool highlights) {
    HairParts parts = create_short(color, length, with_bangs, highlights);
    double tail_w = 16.0;
    double tail_h = std::min(length * 0.7, 60.0);
    parts.emplace_back(
        "rect",
        std::unordered_map<std::string, Value>{{"x", -tail_w / 2.0}, {"y", -k_head_w / 2.0 - 8.0}, {"w", tail_w}, {"h", tail_h}},
        color,
        "#1a1a1a",
        1.5);
    parts.emplace_back(
        "ellipse",
        std::unordered_map<std::string, Value>{{"cx", 0.0}, {"cy", -k_head_w / 2.0 - 8.0}, {"rx", tail_w / 2.0 + 2.0}, {"ry", 4.0}},
        "#c0392b",
        "#1a1a1a",
        1.0);
    return parts;
}

HairParts create_twintails(const std::string& color, double length, bool with_bangs, bool highlights) {
    HairParts parts = create_short(color, length, with_bangs, highlights);
    double tail_h = std::min(length * 0.6, 50.0);
    for (double dx : {-k_head_w / 2.0 - 8.0, k_head_w / 2.0 + 8.0}) {
        parts.emplace_back(
            "rect",
            std::unordered_map<std::string, Value>{{"x", dx - 7.0}, {"y", -k_head_w / 4.0}, {"w", 14.0}, {"h", tail_h}},
            color,
            "#1a1a1a",
            1.5);
        parts.emplace_back(
            "ellipse",
            std::unordered_map<std::string, Value>{{"cx", dx}, {"cy", -k_head_w / 4.0}, {"rx", 9.0}, {"ry", 4.0}},
            "#e74c3c",
            "#1a1a1a",
            1.0);
    }
    return parts;
}

HairParts create_spiky(const std::string& color, double /*length*/, bool with_bangs, bool /*highlights*/) {
    HairParts parts;
    double hw = k_head_w / 2.0 + 6.0;
    int spike_count = 7;
    for (int i = 0; i < spike_count; ++i) {
        double angle = -3.14159265358979 + (3.14159265358979 * i / (spike_count - 1));
        double bx = hw * std::cos(angle) * 0.9;
        double by = hw * std::sin(angle) * 0.9 - 4.0;
        double tx = hw * std::cos(angle) * 1.5;
        double ty = hw * std::sin(angle) * 1.5 - 4.0;
        double perp_x = -std::sin(angle) * 6.0;
        double perp_y = std::cos(angle) * 6.0;
        parts.emplace_back(
            "polygon",
            std::unordered_map<std::string, Value>{
                {"points",
                 poly({bx + perp_x, by + perp_y,
                       tx, ty,
                       bx - perp_x, by - perp_y})}},
            color,
            "#1a1a1a",
            1.5);
    }
    if (with_bangs) {
        auto b = bangs(color);
        parts.insert(parts.end(), b.begin(), b.end());
    }
    return parts;
}

HairParts create_bob(const std::string& color, double length, bool with_bangs, bool /*highlights*/) {
    HairParts parts;
    double cap_h = 20.0;
    parts.emplace_back(
        "ellipse",
        std::unordered_map<std::string, Value>{{"cx", 0.0}, {"cy", -k_head_w / 2.0 + 4.0}, {"rx", k_head_w / 2.0 + 6.0}, {"ry", cap_h}},
        color,
        "#1a1a1a",
        1.5);
    double bob_h = std::min(length * 0.4, 28.0);
    for (double dx : {-k_head_w / 2.0 - 6.0, k_head_w / 2.0 + 6.0}) {
        parts.emplace_back(
            "ellipse",
            std::unordered_map<std::string, Value>{{"cx", dx}, {"cy", 0.0}, {"rx", 10.0}, {"ry", bob_h}},
            color,
            "#1a1a1a",
            1.0);
    }
    if (with_bangs) {
        auto b = bangs(color);
        parts.insert(parts.end(), b.begin(), b.end());
    }
    return parts;
}

HairParts create_bald(const std::string& /*color*/, double /*length*/, bool /*bangs*/, bool /*highlights*/) {
    HairParts parts;
    parts.emplace_back(
        "ellipse",
        std::unordered_map<std::string, Value>{{"cx", -6.0}, {"cy", -k_head_w / 2.0 + 8.0}, {"rx", 8.0}, {"ry", 5.0}},
        "#FFFFFF");
    return parts;
}

using HairMaker = HairParts(*)(const std::string&, double, bool, bool);

const std::unordered_map<std::string, HairMaker> k_hair_makers = {
    {"short", create_short},
    {"medium", create_medium},
    {"long", create_long},
    {"ponytail", create_ponytail},
    {"twintails", create_twintails},
    {"spiky", create_spiky},
    {"bob", create_bob},
    {"braid", create_long},
    {"bun", create_short},
    {"mohawk", create_spiky},
    {"bald", create_bald},
};

}  // anonymous namespace

// ═══════════════════════════════════════════════════════════════════════════════
// Schema
// ═══════════════════════════════════════════════════════════════════════════════

ParamSchema hair_schema() {
    return ParamSchema()
        .enum_("style",
               {"short", "medium", "long", "ponytail", "twintails", "spiky",
                "bob", "braid", "bun", "mohawk", "bald"},
               "medium")
        .color("color", "#2c2c2c")
        .number("length", 60.0, 10.0, 200.0)
        .boolean("bangs", true)
        .boolean("highlights", false);
}

// ═══════════════════════════════════════════════════════════════════════════════
// Factory
// ═══════════════════════════════════════════════════════════════════════════════

std::shared_ptr<GraphicObject> create_hair(
    const std::unordered_map<std::string, Value>& kwargs) {
    auto p = validate_params(hair_schema(), kwargs);

    std::string style = std::get<std::string>(p["style"]);
    std::string color = std::get<std::string>(p["color"]);
    double length = std::get<double>(p["length"]);
    bool bangs_flag = std::get<bool>(p["bangs"]);
    bool highlights = std::get<bool>(p["highlights"]);

    HairMaker maker = create_medium;
    auto it = k_hair_makers.find(style);
    if (it != k_hair_makers.end()) {
        maker = it->second;
    }

    HairParts parts = maker(color, length, bangs_flag, highlights);

    return std::make_shared<GraphicObject>(
        "group",
        std::unordered_map<std::string, Value>{},
        std::nullopt,
        std::nullopt,
        1.0,
        AffineTransform::identity(),
        std::move(parts),
        std::unordered_map<std::string, Value>{
            {"component", std::string("hair")},
            {"style", style},
            {"color", color}});
}

}  // namespace pml

File: pml/sprites/components/hair.py
"""Hair component — generates hairstyle GraphicObjects."""

from __future__ import annotations

from typing import Any

from pml.graphics.objects import GraphicObject
from pml.sprites.validator import ParamSchema, validate_params
from pml.transform import AffineTransform

from .view_utils import VIEW_NAMES, sym_str

_HAIR_SCHEMA = (
    ParamSchema()
    .enum(
        "style",
        ["short", "medium", "long", "ponytail", "twintails", "spiky", "bob", "braid", "bun", "mohawk", "bald"],
        "medium",
    )
    .color("color", "#2c2c2c")
    .number("length", 60, min_val=10, max_val=200)
    .boolean("bangs", True)
    .boolean("highlights", False)
    .enum("view", VIEW_NAMES, "front")
)


def _bangs(color: str, head_w: float) -> list[GraphicObject]:
    """Create bangs (fringe) across the forehead."""
    parts: list[GraphicObject] = []
    bw = head_w * 0.85
    bh = 14
    # Main bangs block
    parts.append(GraphicObject(
        shape_type="rect",
        params={"x": -bw / 2, "y": -head_w / 2 - 4, "w": bw, "h": bh},
        fill=color,
        stroke="#1a1a1a",
        stroke_width=1.5,
    ))
    # Strand lines
    strand_spacing = bw / 6
    for i in range(5):
        sx = -bw / 2 + strand_spacing * (i + 1)
        parts.append(GraphicObject(
            shape_type="line",
            params={"x1": sx, "y1": -head_w / 2 - 4, "x2": sx + 3, "y2": -head_w / 2 - 4 + bh},
            stroke="#1a1a1a",
            stroke_width=0.8,
        ))
    return parts


def _create_short(color: str, head_w: float, length: float, bangs: bool, highlights: bool) -> list[GraphicObject]:
    parts: list[GraphicObject] = []
    # Cap shape on top of head
    cap_h = 16
    parts.append(GraphicObject(
        shape_type="ellipse",
        params={"cx": 0, "cy": -head_w / 2 + 4, "rx": head_w / 2 + 4, "ry": cap_h},
        fill=color,
        stroke="#1a1a1a",
        stroke_width=1.5,
    ))
    # Side tufts
    for dx in (-head_w / 2 - 2, head_w / 2 + 2):
        parts.append(GraphicObject(
            shape_type="ellipse",
            params={"cx": dx, "cy": -head_w / 4, "rx": 6, "ry": 12},
            fill=color,
            stroke="#1a1a1a",
            stroke_width=1.0,
        ))
    if bangs:
        parts.extend(_bangs(color, head_w))
    return parts


def _create_medium(color: str, head_w: float, length: float, bangs: bool, highlights: bool) -> list[GraphicObject]:
    parts: list[GraphicObject] = []
    # Top cap
    parts.append(GraphicObject(
        shape_type="ellipse",
        params={"cx": 0, "cy": -head_w / 2 + 4, "rx": head_w / 2 + 5, "ry": 18},
        fill=color,
        stroke="#1a1a1a",
        stroke_width=1.5,
    ))
    # Side hair flowing down
    side_h = min(length * 0.6, 40)
    for dx in (-head_w / 2 - 4, head_w / 2 + 4):
        parts.append(GraphicObject(
            shape_type="rect",
            params={"x": dx - 8, "y": -head_w / 4, "w": 16, "h": side_h},
            fill=color,
            stroke="#1a1a1a",
            stroke_width=1.0,
        ))
    if bangs:
        parts.extend(_bangs(color, head_w))
    return parts


def _create_long(color: str, head_w: float, length: float, bangs: bool, highlights: bool) -> list[GraphicObject]:
    parts: list[GraphicObject] = []
    # Top cap
    parts.append(GraphicObject(
        shape_type="ellipse",
        params={"cx": 0, "cy": -head_w / 2 + 4, "rx": head_w / 2 + 6, "ry": 20},
        fill=color,
        stroke="#1a1a1a",
        stroke_width=1.5,
    ))
    # Long side hair
    side_h = min(length * 0.8, 80)
    for dx in (-head_w / 2 - 5, head_w / 2 + 5):
        parts.append(GraphicObject(
            shape_type="rect",
            params={"x": dx - 10, "y": -head_w / 4, "w": 20, "h": side_h},
            fill=color,
            stroke="#1a1a1a",
            stroke_width=1.0,
        ))
    # Back hair
    parts.append(GraphicObject(
        shape_type="rect",
        params={"x": -head_w / 2 - 4, "y": -head_w / 2 + 10, "w": head_w + 8, "h": length * 0.7},
        fill=color,
        stroke="#1a1a1a",
        stroke_width=1.0,
    ))
    if bangs:
        parts.extend(_bangs(color, head_w))
    return parts


def _create_ponytail(color: str, head_w: float, length: float, bangs: bool, highlights: bool) -> list[GraphicObject]:
    parts = _create_short(color, head_w, length, bangs, highlights)
    # Ponytail behind head
    tail_w = 16
    tail_h = min(length * 0.7, 60)
    parts.append(GraphicObject(
        shape_type="rect",
        params={"x": -tail_w / 2, "y": -head_w / 2 - 8, "w": tail_w, "h": tail_h},
        fill=color,
        stroke="#1a1a1a",
        stroke_width=1.5,
    ))
    # Tie
    parts.append(GraphicObject(
        shape_type="ellipse",
        params={"cx": 0, "cy": -head_w / 2 - 8, "rx": tail_w / 2 + 2, "ry": 4},
        fill="#c0392b",
        stroke="#1a1a1a",
        stroke_width=1.0,
    ))
    return parts


def _create_twintails(color: str, head_w: float, length: float, bangs: bool, highlights: bool) -> list[GraphicObject]:
    parts = _create_short(color, head_w, length, bangs, highlights)
    tail_h = min(length * 0.6, 50)
    for dx in (-head_w / 2 - 8, head_w / 2 + 8):
        parts.append(GraphicObject(
            shape_type="rect",
            params={"x": dx - 7, "y": -head_w / 4, "w": 14, "h": tail_h},
            fill=color,
            stroke="#1a1a1a",
            stroke_width=1.5,
        ))
        # Tie
        parts.append(GraphicObject(
            shape_type="ellipse",
            params={"cx": dx, "cy": -head_w / 4, "rx": 9, "ry": 4},
            fill="#e74c3c",
            stroke="#1a1a1a",
            stroke_width=1.0,
        ))
    return parts


def _create_spiky(color: str, head_w: float, length: float, bangs: bool, highlights: bool) -> list[GraphicObject]:
    parts: list[GraphicObject] = []
    hw = head_w / 2 + 6
    # Spikes radiating outward
    spike_count = 7
    for i in range(spike_count):
        angle = -3.14 + (3.14 * i / (spike_count - 1))
        import math
        bx = hw * math.cos(angle) * 0.9
        by = hw * math.sin(angle) * 0.9 - 4
        tx = hw * math.cos(angle) * 1.5
        ty = hw * math.sin(angle) * 1.5 - 4
        # Each spike is a triangle
        perp_x = -math.sin(angle) * 6
        perp_y = math.cos(angle) * 6
        points = [
            (bx + perp_x, by + perp_y),
            (tx, ty),
            (bx - perp_x, by - perp_y),
        ]
        parts.append(GraphicObject(
            shape_type="polygon",
            params={"points": points},
            fill=color,
            stroke="#1a1a1a",
            stroke_width=1.5,
        ))
    if bangs:
        parts.extend(_bangs(color, head_w))
    return parts


def _create_bob(color: str, head_w: float, length: float, bangs: bool, highlights: bool) -> list[GraphicObject]:
    parts: list[GraphicObject] = []
    # Rounded cap + sides that curve inward
    cap_h = 20
    parts.append(GraphicObject(
        shape_type="ellipse",
        params={"cx": 0, "cy": -head_w / 2 + 4, "rx": head_w / 2 + 6, "ry": cap_h},
        fill=color,
        stroke="#1a1a1a",
        stroke_width=1.5,
    ))
    # Bob sides (wider, ending at jawline)
    bob_h = min(length * 0.4, 28)
    for dx in (-head_w / 2 - 6, head_w / 2 + 6):
        parts.append(GraphicObject(
            shape_type="ellipse",
            params={"cx": dx, "cy": 0, "rx": 10, "ry": bob_h},
            fill=color,
            stroke="#1a1a1a",
            stroke_width=1.0,
        ))
    if bangs:
        parts.extend(_bangs(color, head_w))
    return parts


def _create_bald(color: str, head_w: float, length: float, bangs: bool, highlights: bool) -> list[GraphicObject]:
    # Bald — just a subtle shine highlight
    return [GraphicObject(
        shape_type="ellipse",
        params={"cx": -6, "cy": -head_w / 2 + 8, "rx": 8, "ry": 5},
        fill="#FFFFFF40",
    )]


_HAIR_MAKERS = {
    "short": _create_short,
    "medium": _create_medium,
    "long": _create_long,
    "ponytail": _create_ponytail,
    "twintails": _create_twintails,
    "spiky": _create_spiky,
    "bob": _create_bob,
    "braid": _create_long,  # fallback to long
    "bun": _create_short,  # fallback to short
    "mohawk": _create_spiky,  # fallback to spiky variant
    "bald": _create_bald,
}


def create_hair(**kwargs: Any) -> GraphicObject:
    """Create a hairstyle graphic object.

    Args:
        :style — 'short | 'medium | 'long | 'ponytail | 'twintails | 'spiky | 'bob | ...
        :color — hair color
        :length — hair length in pixels (clamped 10-200)
        :bangs — whether to include bangs
        :highlights — whether to add highlight strands
        :view — 'front | 'side | 'back | 'three-quarter
    """
    p = validate_params(_HAIR_SCHEMA, {sym_str(k): v for k, v in kwargs.items()})

    style = p["style"]
    color = p["color"]
    length = p["length"]
    bangs = p["bangs"]
    highlights = p["highlights"]
    view = p["view"]

    head_w = 56

    if view == "back":
        # No bangs from behind, same head width
        bangs = False
    elif view == "side":
        head_w = int(head_w * 0.5)
    elif view == "three-quarter":
        head_w = int(head_w * 0.75)

    maker = _HAIR_MAKERS.get(style, _create_medium)
    parts = maker(color, head_w, length, bangs, highlights)

    # Side view: shift hair to the right
    if view == "side":
        for i, part in enumerate(parts):
            parts[i] = part.with_transform(
                AffineTransform.translate(12, 0)
            )

    return GraphicObject(
        shape_type="group",
        children=tuple(parts),
        metadata={"component": "hair", "style": style, "color": color, "view": view},
    )
